UniversalAlgebra/ClosureSystems/Basic.lagda

```diff
@@ -34,30 +34,26 @@ Some examples of closure systems are the following:
 
 module ClosureSystems.Basic where
 
-open import Agda.Primitive             using    ( _⊔_   ;  Level
-                                                ; lsuc            )
-                                       renaming ( Set   to Type   )
-open import Data.Empty.Polymorphic     using    ( ⊥ )
-open import Data.Unit.Polymorphic      using    (⊤ ; tt)
-open import Relation.Binary.Bundles    using    ( Poset           )
-open import Relation.Unary             using    ( Pred  ;   _⊆_
-                                                ; _∈_   ;   ⋂     )
-
-open import ClosureSystems.Definitions using    ( Extensive ; OrderPreserving ; Idempotent )
-
--- universe-polymorphic emptyset type
-∅ : {ℓ ℓ₁ : Level}{A : Type ℓ} → Pred A ℓ₁
-∅ = λ _ → ⊥
-
--- closure system
-data 𝒞𝓁 {ℓ : Level}(X : Type ℓ) : Pred (Pred X ℓ) (ℓ ⊔ lsuc ℓ) where
- nul : ∅ ∈ 𝒞𝓁 X
- all  : (λ _ → ⊤) ∈ 𝒞𝓁 X
- and : {I : Type}(c : I → Pred X ℓ) → (∀ i → c i ∈ 𝒞𝓁 X) → (⋂ I c) ∈ 𝒞𝓁 X
+open import Agda.Primitive           using ( _⊔_ ; lsuc )     renaming ( Set to Type )
+import Algebra.Definitions
+open import Data.Product             using ( Σ-syntax )
+open import Level                    using ( Level ; Lift )   renaming ( zero to ℓ₀ )
+open import Relation.Binary.Bundles  using ( Poset )
+open import Relation.Binary.Core     using ( _Preserves_⟶_ )
+open import Relation.Unary           using ( Pred ; _∈_ ; ⋂ )
+
+open import ClosureSystems.Definitions using ( Extensive )
 
-\end{code}
 
+module _ {χ ℓ ρ ι : Level}{X : Type χ} where
 
+ IntersectClosed : Pred (Pred X (ℓ ⊔ ι) ) ρ → Type _
```

Comment:
Having a weird Level like `ℓ ⊔ ι` outside of what is inferred for a Type (instead of a simple variable) is an Agda "design smell". You're shifting level problems around. Definitions should have simple types, and the use sites might have level restrictions because they aren't valid in full generality.  The `Lift` in the definition is another sign.

Reply:
Yes, I thought this smelled funky... that's why I posted a [message on zulip about this](https://agda.zulipchat.com/#narrow/stream/259644-newcomers/topic/Closure.20system.20as.20inductive.20data.20type/near/246095908).  Did you see it?

I don't know what a better alternative would be, but I will look at it again now.

I ran into the same problem (re. levels) when defining the inductive types of closure operators H, S, P, and V.  Those are especially funky.  I redefined them all at least a hundred times to get versions that work well enough to do what I needed them to.  I thought they were getting close to the "right" definitions, but then you improved them substantially by removing unnecessary constructors.  That was great!

I'm still unsure if the way H, S, P, and V use levels is the best possible... but at least these types are now flexible enough to prove things that I couldn't prove with all previous versions.